### ISSUE TITLE
Deeplab V3+ and Xception

### ISSUE THREAD

**Issue by lromor** [yassouali/pytorch_segmentation#78]
Hi!
Great repo.
Could you recommend a configuration file for running an experiment using Deeplab V3+ and Xception
to achieve at some level similar to the results presented in the paper https://arxiv.org/pdf/1802.02611.pdf?

I'm constantly getting very low mIOUs with the following:

```
{
    "name": "DeepLab",
    "n_gpu": 1,
    "use_synch_bn": true,

    "arch": {
        "type": "DeepLab",
        "args": {
            "backbone": "xception",
            "freeze_bn": false,
            "freeze_backbone": false
        }
    },

    "train_loader": {
        "type": "VOC",
        "args":{
            "data_dir": "data/VOCtrainval_11-May-2012",
            "batch_size": 8,
            "crop_size": 513,
            "augment": true,
            "shuffle": true,
            "scale": true,
            "flip": true,
            "rotate": false,
            "blur": false,
            "split": "train_aug",
            "num_workers": 4
        }
    },

    "val_loader": {
        "type": "VOC",
        "args":{
            "data_dir": "data/VOCtrainval_11-May-2012",
            "batch_size": 8,
            "crop_size": 513,
            "val": true,
            "split": "val",
            "num_workers": 4
        }
    },

    "optimizer": {
        "type": "SGD",
        "differential_lr": true,
        "args":{
            "lr": 0.01,
            "weight_decay": 1e-4,
            "momentum": 0.9
        }
    },

    "loss": "CrossEntropyLoss2d",
    "ignore_index": 255,
    "lr_scheduler": {
        "type": "Poly",
        "args": {}
    },

    "trainer": {
        "epochs": 80,
        "save_dir": "saved/",
        "save_period": 10,
  
        "monitor": "max Mean_IoU",
        "early_stop": 10,
        
        "tensorboard": true,
        "log_dir": "saved/runs",
        "log_per_iter": 20,

        "val": true,
        "val_per_epochs": 5
    }
}

```

**Comment by yassouali** [comment]
Hi @lromor,

For deeplab v3+ with xception backbone, the backbone used is not really the same, if you go through the code, you'll see that the checkpoint model we're using from [pretrained-models.pytorch](https://github.com/Cadene/pretrained-models.pytorch) is a smaller version that the one deeplab v3+ uses, and the layers not in the checkpoint are initialized using the last layer in the checkpoint. I think this is why you might have some problems when training. 

I suggest to not use differentiable learning rate (where the backbone is trained with 0.1lr), and use the same learning rate across the whole model to train the backbone too. 

If you like, I am sure there is some ported deeplab xception checkpoint you can find, and in this case, you can load the correct weights and train normally (I might do this if I have sometime).

**Comment by lromor** [comment]
I see, thank you for you answer. I'm also noticing another strange behavior with resnet, the validation loss tends to go up after 600 iterations.

**Comment by yassouali** [comment]
Not really. The validation loss might to go up from time to time, but as a general trend, I'd expect it to go down. Maybe try a lower LR.

**Comment by lromor** [comment]
I'm having the very same diverging validation loss of #29 .

**Comment by lromor** [comment]
I'm sorry, I'm relatively new to segmentation, is it normal for the targets
to be tensors of size `torch.Size([N, 96, 96])` where N is the batch size?
The output of the model is correctly [N, C, W, H] but arent' the targets supposed also to be channel wise for each class?
The current target seems just to be an integer tensor with different integer values probably mapping to the different classes.
I don't expect the loss function to work on those.

**Comment by lromor** [comment]
I forgot that CrossEntropyLoss wants labels as integers. Sorry :)

**Comment by lromor** [comment]
Hi @yassouali , I think there's a problem with the configuration options for pascal voc.
The transformations that should be configurable are not actually taking place at all. Can you also confirm?

**Comment by lromor** [comment]
@yassouali it seems to me that by default, the validation set performs scaling augmentation.
So the configuration file you posted might 
```

    "val_loader": {
        "type": "VOC",
        "args":{
            "data_dir": "data/VOCtrainval_11-May-2012",
            "batch_size": 8,
            "crop_size": 513,
            "val": true,
            "split": "val",
            "num_workers": 4
        }
    },
```

this might actually enable scale=true.

**Comment by yassouali** [comment]
@lromor the current config applies a batch validation, and for validation, in base_dataset, the augmentations used in validation are different. We only apply center crops, where the smaller side is automatically rescaled to the crop size. If you want, to get more precise results, you can remove `crop_size` and use a batch of 1. This way you'll be sure to use the original images, but it'll take a bit longer.

**Comment by lromor** [comment]
You are right. I just noticed that the validation does skip the augmentations.

I'm searching for possible reasons. I did update the resnet101 backbone and achieved better results, but just1,2% reaching ~75% on the validation dataset.
One thing I want to ask you, it seems that no matter what, after 5 epochs, the validation loss starts (slowly) increasing.
Do you think this is expected? The mIOU and other metrics do improve over time in the validation set, but 
it's somewhat counterintuitive for the validation to go (slightly but steady) upwards. 

It's similar to what's happening in the plot you can find in the bottom of this page:
https://www.tensorflow.org/tutorials/images/segmentation

One last thing. The original paper uses multi-grid. I'm not sure what does that means exactly, but maybe this implementation lacks of this method leading to lower accuracies. 
What do you think? Otherwise I can't think of other issues.

**Comment by yassouali** [comment]
I think the multi-grid refers to using different `assp_branch` on top of the backbone with different dilation rates. As for the mIoU difference (I see that the expected result with a resnet101 is 78%), as you said, I certainly think the batch norm plays some role in this, fine-tuning batch norm is always tricky, maybe try different lr rates for the batch norm, currently, the backbone is fine tunned with LR/10, maybe you can try the same LR for only batch norm layers, or as you said, only fine-tuning the batch norm layers and freezing the backbone, I think torchvision have a reference implementation of deeplab v3+ if you want to take a look.

And ofcourse, please make a pull request with your findings and correction to the code if you find any improvements.

**Comment by lromor** [comment]
@yassouali , this is a run using the torchvision backbone.
I still used your ASPP implementation and decoder.
This is the problem I was talking about:
![image](https://user-images.githubusercontent.com/1597330/89356043-1a804b80-d6bd-11ea-838f-33a811a6689a.png)

The training goes down, the val loss, goes up.
Notice that even if the trend of the val loss is to go up from the start, all the initial part (first 5 epochs) is missing, and obviously when the model is just initialized the validation loss is much much higher.


The upward trend it's still relatively small, but the interesting thing is that the mIOU keeps improving, as well as the class accuracy.

I wonder how's that possible.

**Comment by yassouali** [comment]
@lromor 

Sorry for the delay,

Yeah, the behavior is certainly interesting, but maybe since the backbone is pretrained, we are already close to a local minimum, and the model only needs to search within the vicinity for this minimum for the optimal performances. Which might explain why the loss remains relatively small, affirming that no overfitting is taking place.

**Comment by lromor** [comment]
I see. Thanks for your help. In my fork I have an updated version with newer backbones. But, as you can see, they are not really changing the results too much.

Regarding the increasing loss:

That makes sense. I thought strange for the validation mIoU and loss to have diverging trends.
